src/plugins/omemo-views/utils.js

```diff
@@ -0,0 +1,154 @@
+/* global libsignal */
+import concat from 'lodash-es/concat';
+import difference from 'lodash-es/difference';
+import log from '@converse/headless/log';
+import tpl_audio from 'templates/audio.js';
+import tpl_file from 'templates/file.js';
+import tpl_image from 'templates/image.js';
+import tpl_video from 'templates/video.js';
+import { __ } from 'i18n';
+import { _converse, converse, api } from '@converse/headless/core';
+import { html } from 'lit';
+import { isError } from '@converse/headless/utils/core.js';
+import { isAudioURL, isImageURL, isVideoURL, getURI } from '@converse/headless/utils/url.js';
+import { until } from 'lit/directives/until.js';
+
+const { Strophe, URI, sizzle, u } = converse.env;
+
+function getTemplateForObjectURL (uri, obj_url, richtext) {
+    if (isError(obj_url)) {
+        return html`<p class="error">${obj_url.message}</p>`;
+    }
+
+    const file_url = uri.toString();
+    if (isImageURL(file_url)) {
+        return tpl_image({
+            'src': obj_url,
+            'onClick': richtext.onImgClick,
+            'onLoad': richtext.onImgLoad
+        });
+    } else if (isAudioURL(file_url)) {
+        return tpl_audio(obj_url);
+    } else if (isVideoURL(file_url)) {
+        return tpl_video(obj_url);
+    } else {
+        return tpl_file(obj_url, uri.filename());
+    }
+
+}
+
+// TODO: move the non-template part to headless
+function addEncryptedFiles(text, offset, richtext) {
+    const objs = [];
+    try {
+        const parse_options = { 'start': /\b(aesgcm:\/\/)/gi };
+        URI.withinString(
+            text,
+            (url, start, end) => {
+                objs.push({ url, start, end });
+                return url;
+            },
+            parse_options
+        );
+    } catch (error) {
+        log.debug(error);
+        return;
+    }
+    objs.forEach(o => {
+        const uri = getURI(text.slice(o.start, o.end));
+        const promise = getAndDecryptFile(uri)
+            .then(obj_url => getTemplateForObjectURL(uri, obj_url, richtext));
+
+        const template = html`${until(promise, '')}`;
```

Comment:
@jcbrand this function is stuck in `omemo-views` only because of the last bit (`getTemplateForObjectURL` and render). Any ideas on how to make this a separate function/call and have the rest in headless?

Possible solution: I could have `[headless].addEncryptedFiles()` which returns a promise, and a `[views].addEncryptedFiles()` which calls `[headless].addEncryptedFiles()` but then listens to the promise and renders the template once it's done. (I'm a little leery of returning promises to function calls because I haven't done it much before, but that might be personal bias)

Reply:
Return a list of promises (with each promise resolving into an encrypted file's URL).

Then in the view-code, loop through the list of promises and for each call `getTemplateForObjectURL`, but the result in a template like on line 62 and then call `richtext.addTemplateResult` with that template.

Reply:
> I'm a little leery of returning promises to function calls because I haven't done it much before, but that might be personal bias

It's totally normal and just something to get used to :)
Just remember to either `await` or the chain with `.then()`.

Reply:
I ended up having to return them as one of many properties, which may not be ideal. Anyway, will flag it in a fresh thread. Thanks for the reassurance about returning promises :sweat_smile: